<!--
  ~ Copyright (c) 2003-2022 Regents of the University of California and Broad Institute. All rights reserved.
  -->

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
    <head>
        <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1" />
        <META HTTP-EQUIV="CACHE-CONTROL" CONTENT="NO-CACHE" />
        <title>Differential Expression Analysis</title>
        <link href="../../css/protocolsWindow.css" rel="stylesheet" type="text/css" />
        <script type="text/javascript" src="../../js/protocol.js"></script>
    </head>

    <body>
        <h1>Differential Expression Analysis</h1>
        <table width="100%" border="0" cellpadding="0" cellspacing="0">
            <tr>
                <td align="left">
                    <a href="protocols.html">protocols</a>
                </td>
            </tr>
        </table>
        <p>Find genes that are significantly differentially expressed between classes of samples.</p>

        <div class="steps">
            <h2>Before you begin</h2>
            <ul>
                <li>Gene expression data must be in a <a href="GctResFiles.html">GCT or RES file</a>.
                    <br />Example file: <a href="//software.broadinstitute.org/cancer/software/genepattern/data/all_aml/all_aml_test.gct">all_aml_test.gct</a>.
                </li>
                <li>The class of each sample must be identified in a <a href="ClsFiles.html">CLS file</a>.
                    <br />Example file: <a href="//software.broadinstitute.org/cancer/software/genepattern/data/all_aml/all_aml_test.cls">all_aml_test.cls</a>.
                </li>
            </ul>
            <table width="100%" border="0" cellpadding="0" cellspacing="0">
                <tr>
                    <td align="right">learn more:<br />
                        <a href="http://software.broadinstitute.org/cancer/software/genepattern/file-formats-guide">file formats</a>
                    </td>
                </tr>
            </table>

        </div>
        <div class="steps">
            <h2>Step 1: PreprocessDataset</h2>

            <p>Preprocess gene expression data
                to remove platform noise and genes that have little variation.</p>


            <div class="gpLinks">
                <ul>
                    <li><a href="/gp/pages/index.jsf?lsid=PreprocessDataset"
                           onclick="return openModuleWindow(this, event);"
                            >Open module</a> in the GenePattern window.
                    </li>
                    <li><a href="/gp/pages/index.jsf?lsid=PreprocessDataset&input.filename=https://software.broadinstitute.org/cancer/software/genepattern/data/all_aml/all_aml_test.gct"
                           onclick="return openModuleWindow(this, event);"
                            >Open module with example data</a> in the GenePattern window.
                    </li>
                </ul>
            </div>
            <h5>Considerations</h5>
            <ul>
                <li>PreprocessDataset can preprocess the data in one or more ways (in this order):
                    <ol>
                        <li>Set threshold and ceiling values. Any value lower/higher than the threshold/ceiling
                            value is reset to the threshold/ceiling value.
                        </li>
                        <li>Convert each expression value to the log base 2 of the value.</li>
                        <li>Remove genes (rows) if a given number of its sample values are less than
                            a given threshold.
                        </li>
                        <li>Remove genes (rows) that do not have a minimum fold change or expression
                            variation.
                        </li>
                        <li>Discretize or normalize the data.</li>
                    </ol>
                </li>

                <li>ComparativeMarkerSelection expects non-log-transformed data. Some calculations, such as Fold Change, will produce incorrect results on log transformed data.</li>
                <li>If you did not generate the expression data,
                    check whether preprocessing steps have already been taken before
                    running the PreprocessDataset module.
                </li>
            </ul>


            <table width="100%" border="0" cellpadding="0" cellspacing="0">
                <tr>
                    <td align="right">learn more:<br />
                        <a href="/gp/getTaskDoc.jsp?name=PreprocessDataset">PreprocessDataset</a>
                    </td>
                </tr>
            </table>
        </div>

        </div>

        <div class="steps">
            <h2>Step 2: ComparativeMarkerSelection</h2>

            <p>ComparativeMarkerSelection computes differential gene expression.
                For each gene, it uses a test statistic to calculate the
                difference in gene expression between classes and then computes a p-value to estimate the
                significance of the test statistic score. </p>

            <p>Because testing tens of thousands of genes simultaneously
                increases the possibility of mistakenly identifying a non-marker gene as a marker gene (a false positive),
                ComparativeMarkerSelection corrects for multiple hypothesis testing by computing both false discovery
                rates (FDR) and family-wise error rates (FWER).
            </p>


            <div class="gpLinks">
                <ul>
                    <li><a href="/gp/pages/index.jsf?lsid=ComparativeMarkerSelection"
                           onclick="return openModuleWindow(this, event);"
                            >Open module</a></li>
                    <li><a href="/gp/pages/index.jsf?lsid=ComparativeMarkerSelection&input.file=https://software.broadinstitute.org/cancer/software/genepattern/data/protocols/all_aml_test.preprocessed.gct&cls.file=https://software.broadinstitute.org/cancer/software/genepattern/data/all_aml/all_aml_test.cls"
                           onclick="return openModuleWindow(this, event);"
                            >Open module with example data</a>
                    </li>
                </ul>
            </div>

            <h5>Considerations</h5>
            <ul>
                <li>If the data set includes at least 10 samples per class, use the default value of 1000 permutations
                    to ensure accurate <i>p</i>-values. If the data set includes fewer than 10 samples in any class, permuting the
                    samples cannot give an accurate <i>p</i>-value; specify 0 permutations to use asymptotic <i>p</i>-values
                    instead.
                </li>
                <li>If the data set includes more than two classes, use the <i>phenotype test</i> parameter
                    to analyze each class against all others (one-versus-all) or all class pairs
                    (all pairs).
                </li>
            </ul>


            <table width="100%" border="0" cellpadding="0" cellspacing="0">
                <tr>
                    <td align="right">learn more:<br />
                        <a href="/gp/getTaskDoc.jsp?name=ComparativeMarkerSelection">ComparativeMarkerSelection</a>
                    </td>
                </tr>
            </table>
        </div>


        <div class="steps">
            <h2>Step 3: ComparativeMarkerSelectionViewer</h2>

            <p>Run the ComparativeMarkerSelectionViewer module to view the results.
                The viewer displays the test statistic score, its p value, two FDR statistics and three FWER statistics for each gene.
            </p>

            <div class="gpLinks">
                <ul>
                    <li><a href="/gp/pages/index.jsf?lsid=ComparativeMarkerSelectionViewer"
                           onclick="return openModuleWindow(this, event);"
                            >Open module</a></li>
                    <li><a href="/gp/pages/index.jsf?lsid=ComparativeMarkerSelectionViewer&comparative.marker.selection.filename=https://software.broadinstitute.org/cancer/software/genepattern/data/protocols/all_aml_test.preprocessed.comp.marker.odf&dataset.filename=https://software.broadinstitute.org/cancer/software/genepattern/data/protocols/all_aml_test.preprocessed.gct"
                           onclick="return openModuleWindow(this, event);"
                            >Open module with example data</a></li>
                </ul>
            </div>
            <h5>Considerations</h5>
            <ul>
                <li>Generally,
                    researchers identify marker genes based on FDR rather than the more conservative FWER.
                </li>
                <li>Often, marker genes are identified based on an FDR cutoff value of .05, which
                    indicates that a gene identified as a marker gene has a 1 in 20 (5%) chance of being a false positive.
                    Select <i>Edit&gt;Filter Features&gt;Custom Filter</i> to filter results
                    based on that criteria (or any other).
                </li>
                <li>Select <i>File&gt;Save Derived Dataset</i> to create a GCT file that contains
                    a subset of the expression data.
                </li>
            </ul>


            <table width="100%" border="0" cellpadding="0" cellspacing="0">
                <tr>
                    <td align="right">learn more:<br />
                        <a href="/gp/getTaskDoc.jsp?name=ComparativeMarkerSelectionViewer">ComparativeMarkerSelectionViewer</a>
                    </td>
                </tr>
            </table>


        </div>

        <h2>Reference</h2>

        <p>Gould, J., Getz, G., Monti, S., Reich, M., and Mesirov, J.P. 2006.
            Comparative gene marker selection suite. <i>Bioinformatics</i> 22(15):1924-1925.</p>

    </body>
</html>
