<div id="runTaskSettingsDiv">
                <div class="task_header" id="taskHeaderDiv">
                    <span id="task_name" class="module_header ">CufflinksCuffmergePipeline</span>
                <span id="task_version_main">
                    <label for="task_versions">
                        version
                    </label>
                    <span class="normal">1</span>
                </span>

                    <div id="otherOptionsDiv" class="floatRight">
                        <img id="otherOptions" alt="other options" src="/gp/images/gear_blue_and_white.png" height="20">
                            <ul id="otherOptionsMenu" class="module-menu ui-menu ui-widget ui-widget-content ui-corner-all" style="position: absolute; top: 22px; left: -164px; display: none;" role="menu" tabindex="0" aria-activedescendant="properties">
                                <li class="ui-menu-item" role="presentation">
                                    <a id="properties" href="#" onclick="jq('#otherOptionsMenu').hide();showProperties();" class="ui-corner-all" tabindex="-1" role="menuitem">Properties</a>
                                </li>
                                <li class="ui-menu-item" role="presentation">
                                    <a id="clone" href="#" onclick="jq('#otherOptionsMenu').hide();cloneTask();" class="ui-corner-all" tabindex="-1" role="menuitem">Clone</a>
                                </li>
                                <li class="ui-menu-item" role="presentation">
                                    <a id="export" target="_blank" onclick="jq('#otherOptionsMenu').hide();" href="/gp/makeZip.jsp?name=urn:lsid:broad.mit.edu:cancer.software.genepattern.module.pipeline:00032:1" class="ui-corner-all" tabindex="-1" role="menuitem">Export</a>
                                </li>
                                <li class="ui-menu-item" role="presentation">
                                    <a id="javaCode" class="viewCode ui-corner-all" href="#" onclick="jq('#otherOptionsMenu').hide();" tabindex="-1" role="menuitem">Java code</a>
                                </li>
                                <li class="ui-menu-item" role="presentation">
                                    <a id="matlabCode" class="viewCode ui-corner-all" href="#" onclick="jq('#otherOptionsMenu').hide();" tabindex="-1" role="menuitem">MATLAB code</a>
                                </li>
                                <li class="ui-menu-item" role="presentation">
                                    <a id="rCode" class="viewCode ui-corner-all" href="#" onclick="jq('#otherOptionsMenu').hide();" tabindex="-1" role="menuitem">R code</a>
                                </li>
                                <li class="ui-menu-item" role="presentation">
                                    <a onclick="jq('#otherOptionsMenu').hide();" id="ui-id-2134" class="ui-corner-all" tabindex="-1" role="menuitem">
                                    <input name="toggleDesc" id="toggleDesc" type="checkbox" checked="checked">
                                    <label for="toggleDesc">
                                        Show parameter descriptions
                                    </label>
                                    </a>
                                </li>
                            </ul>
                    </div>
                    <a class="floatRight" id="documentation" href="/gp/getTaskDoc.jsp?name=urn:lsid:broad.mit.edu:cancer.software.genepattern.module.pipeline:00032:1" target="_blank">Documentation</a>

                </div>
                <div class="clear" id="viewCodeDiv" style="display: none;">
                    <img class="floatRight ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" alt="close view code text" id="removeViewCode" src="/gp/images/delete-blue.png" role="button" aria-disabled="false"></img>
                </div>

                <div class="clear propertiesContent solid-border paramgroup-spacing" id="viewProperties" style="">
                    <img class="floatRight ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" alt="close view code text" id="removeViewProperties" src="/gp/images/delete-blue.png" role="button" aria-disabled="false"></img>
                    <table>
                        <tbody><tr>
                            <td class="properties-label">Name</td>
                            <td class="properties-name">CufflinksCuffmergePipeline</td>
                        </tr>
                        <tr>
                            <td class="properties-label">LSID</td>
                            <td class="properties-lsid">urn:lsid:broad.mit.edu:cancer.software.genepattern.module.pipeline:00032:1</td>
                        </tr>
                        <tr>
                            <td class="properties-label">Description</td>
                            <td class="properties-description">Creates multiple transcript assemblies using Cufflinks and then merges these assemblies into one</td>
                        </tr>
                        <tr>
                            <td class="properties-label">Author</td>
                            <td class="properties-author"></td>
                        </tr>
                        <tr>
                            <td class="properties-label">Privacy</td>
                            <td class="properties-privacy">public</td>
                        </tr>
                        <tr>
                            <td class="properties-label">License</td>
                            <td class="properties-license">None</td>
                        </tr>
                        <tr>
                            <td class="properties-label">Quality Level</td>
                            <td class="properties-quality">development</td>
                        </tr>
                        <tr>
                            <td class="properties-label">Documentation</td>
                            <td class="properties-documentation"><a href="/gp/getTaskDoc.jsp?name=urn:lsid:broad.mit.edu:cancer.software.genepattern.module.pipeline:00032:1">Click Here</a></td>
                        </tr>
                        <tr>
                            <td class="properties-label">Command Line</td>
                            <td class="properties-commandline">&lt;java&gt; -cp &lt;pipeline.cp&gt; -Dgenepattern.properties=&lt;resources&gt; -DLSID=&lt;LSID&gt; -DPIPELINE_ARG.StopAfterTask=&lt;PIPELINE_ARG.StopAfterTask&gt; -DGenePatternURL=&lt;GenePatternURL&gt;  &lt;pipeline.main&gt; &lt;GenePatternURL&gt;getPipelineModel.jsp?name=&lt;LSID&gt;&amp;userid=&lt;userid&gt; &lt;userid&gt; CreateSymlinks1.input.files=&lt;CreateSymlinks1.input.files&gt; CufflinksWrapperPipeline2.Cufflinks1.GTF.guide=&lt;CufflinksWrapperPipeline2.Cufflinks1.GTF.guide&gt; CufflinksWrapperPipeline2.Cufflinks1.mask.file=&lt;CufflinksWrapperPipeline2.Cufflinks1.mask.file&gt; CufflinksWrapperPipeline2.Cufflinks1.library.type=&lt;CufflinksWrapperPipeline2.Cufflinks1.library.type&gt; CufflinksWrapperPipeline2.Cufflinks1.min.frags.per.transfrag=&lt;CufflinksWrapperPipeline2.Cufflinks1.min.frags.per.transfrag&gt; CufflinksWrapperPipeline2.Cufflinks1.additional.cufflinks.options=&lt;CufflinksWrapperPipeline2.Cufflinks1.additional.cufflinks.options&gt; Cufflinks.cuffmerge4.reference.GTF=&lt;Cufflinks.cuffmerge4.reference.GTF&gt; Cufflinks.cuffmerge4.genome.file=&lt;Cufflinks.cuffmerge4.genome.file&gt;</td>
                        </tr>
                        <tr>
                            <td class="properties-label">Categories</td>
                            <td class="properties-tasktype">pipeline;pipeline</td>
                        </tr>
                        <tr>
                            <td class="properties-label">CPU Type</td>
                            <td class="properties-cpu">any</td>
                        </tr>
                        <tr>
                            <td class="properties-label">Operating System</td>
                            <td class="properties-os">any</td>
                        </tr>
                        <tr>
                            <td class="properties-label">Language</td>
                            <td class="properties-language">Java</td>
                        </tr>
                        <tr>
                            <td class="properties-label">Current Files</td>
                            <td class="properties-files"><a href="/gp/getFile.jsp?task=urn%3Alsid%3Abroad.mit.edu%3Acancer.software.genepattern.module.pipeline%3A00032%3A1&amp;file=doc.html">doc.html</a>, <a href="/gp/getFile.jsp?task=urn%3Alsid%3Abroad.mit.edu%3Acancer.software.genepattern.module.pipeline%3A00032%3A1&amp;file=paramgroups.json">paramgroups.json</a>, <a href="/gp/getFile.jsp?task=urn%3Alsid%3Abroad.mit.edu%3Acancer.software.genepattern.module.pipeline%3A00032%3A1&amp;file=manifest">manifest</a>, </td>
                        </tr>
                        <tr>
                            <td class="properties-label">Version Comment</td>
                            <td class="properties-versioncomment"></td>
                        </tr>
                        <tr>
                            <td class="properties-label">File Formats</td>
                            <td class="properties-formats"></td>
                        </tr>
                    </tbody></table>
                    <div class="properties-children"><div><div class="pHeaderTitleDiv background" style="margin-top: 10px;"><img class="paramSectionToggle" height="19" width="19" src="/gp/images/toggle_expand.png">1. <a href="/gp/pages/index.jsf?lsid=urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00343:1" onclick="event.stopPropagation();">CreateSymlinks</a><div style="float: right; font-size: 0.75em; padding-right: 5px;">Version 1</div></div><div class="paramGroupSection dotted-border" style="display: none;">Creates symlinks to the input files in the job results directory. This is intended to be a helper module for creating scatter-gather pipelines.  Note that this module is not supported on Windows<table class="paramGroupSection dotted-border"><tbody><tr><td style="padding-right: 30px; font-weight: bold;">input.files</td><td>Prompt When Run</td></tr></tbody></table></div></div><div><div class="pHeaderTitleDiv background" style="margin-top: 10px;"><img class="paramSectionToggle" height="19" width="19" src="/gp/images/toggle_expand.png">2. <a href="/gp/pages/index.jsf?lsid=urn:lsid:broad.mit.edu:cancer.software.genepattern.module.pipeline:00033:1" onclick="event.stopPropagation();">CufflinksWrapperPipeline</a><div style="float: right; font-size: 0.75em; padding-right: 5px;">Version 1</div></div><div class="paramGroupSection dotted-border" style="display: none;">A wrapper pipeline containing Cufflinks that is meant to be used as part of a scatter-gather pipeline<table class="paramGroupSection dotted-border"><tbody><tr><td style="padding-right: 30px; font-weight: bold;">Cufflinks1.input.file</td><td></td></tr><tr><td style="padding-right: 30px; font-weight: bold;">Cufflinks1.transfrag.label</td><td></td></tr><tr><td style="padding-right: 30px; font-weight: bold;">Cufflinks1.GTF</td><td></td></tr><tr><td style="padding-right: 30px; font-weight: bold;">Cufflinks1.GTF.guide</td><td>Prompt When Run</td></tr><tr><td style="padding-right: 30px; font-weight: bold;">Cufflinks1.mask.file</td><td>Prompt When Run</td></tr><tr><td style="padding-right: 30px; font-weight: bold;">Cufflinks1.frag.bias.correct</td><td></td></tr><tr><td style="padding-right: 30px; font-weight: bold;">Cufflinks1.multi.read.correct</td><td>false</td></tr><tr><td style="padding-right: 30px; font-weight: bold;">Cufflinks1.library.type</td><td>Prompt When Run</td></tr><tr><td style="padding-right: 30px; font-weight: bold;">Cufflinks1.min.frags.per.transfrag</td><td>Prompt When Run</td></tr><tr><td style="padding-right: 30px; font-weight: bold;">Cufflinks1.additional.cufflinks.options</td><td>Prompt When Run</td></tr></tbody></table></div></div><div><div class="pHeaderTitleDiv background" style="margin-top: 10px;"><img class="paramSectionToggle" height="19" width="19" src="/gp/images/toggle_expand.png">3. <a href="/gp/pages/index.jsf?lsid=urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00280:1" onclick="event.stopPropagation();">FilterFilelist</a><div style="float: right; font-size: 0.75em; padding-right: 5px;">Version 1</div></div><div class="paramGroupSection dotted-border" style="display: none;">a helper module which generates a filtered filelist from the input filelist and some filter parameters.<table class="paramGroupSection dotted-border"><tbody><tr><td style="padding-right: 30px; font-weight: bold;">input.filelist</td><td></td></tr><tr><td style="padding-right: 30px; font-weight: bold;">includePattern</td><td>transcripts.gtf</td></tr><tr><td style="padding-right: 30px; font-weight: bold;">type</td><td>f</td></tr><tr><td style="padding-right: 30px; font-weight: bold;">errorOnEmptySet</td><td>yes</td></tr><tr><td style="padding-right: 30px; font-weight: bold;">output.filename</td><td>&lt;input.filelist_basename&gt;.filtered.&lt;input.filelist_extension&gt;</td></tr></tbody></table></div></div><div><div class="pHeaderTitleDiv background" style="margin-top: 10px;"><img class="paramSectionToggle" height="19" width="19" src="/gp/images/toggle_expand.png">4. <a href="/gp/pages/index.jsf?lsid=urn:lsid:broad.mit.edu:cancer.software.genepattern.module.analysis:00291:3" onclick="event.stopPropagation();">Cuffmerge</a><div style="float: right; font-size: 0.75em; padding-right: 5px;">Version 3</div></div><div class="paramGroupSection dotted-border" style="display: none;">Cufflinks 2.0.2: Cuffmerge - Merge several Cufflinks assemblies<table class="paramGroupSection dotted-border"><tbody><tr><td style="padding-right: 30px; font-weight: bold;">input.file</td><td></td></tr><tr><td style="padding-right: 30px; font-weight: bold;">reference.GTF</td><td>Prompt When Run</td></tr><tr><td style="padding-right: 30px; font-weight: bold;">genome.file</td><td>Prompt When Run</td></tr></tbody></table></div></div></div>
                </div>

                <table id="description_main">
                    <tbody>
                    <tr>
                        <td id="mod_description">

                        Creates multiple transcript assemblies using Cufflinks and then merges these assemblies into one</td>
                        <td id="source_info">
                            <div id="source_info_tooltip" style="display: none;"></div>
                        </td>
                    </tr>
                    </tbody>
                </table>

                <hr class="clear">

                <div>
                <span class="otherControlsDiv">
                    * required field
                </span>
                <span class="submitControlsDiv floatRight">
                    <button class="Reset ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" role="button" aria-disabled="false"><span class="ui-button-text"><img src="/gp/images/reset.gif" width="16" height="16" alt="Reset" class="buttonIcon"> Reset </span></button>
                    <button class="Run ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" role="button" aria-disabled="false"><span class="ui-button-text"><img src="/gp/images/run-green.gif" width="16" height="16" alt="Run" class="buttonIcon"> Run </span></button>
                </span>
                </div>

                <hr class="clear">
                <!--<table>
                    <tr>
                        <td><b>Notes:</b> </td>
                        <td>
                            <textarea rows="4" cols="50"> This run tests running CMS using SNR test statistic. </textarea>
                        </td>
                    </tr>
                </table> -->

                <div id="missingTasksDiv"></div>
                <div id="paramsListingDiv">
                    <form id="runTaskForm" action="#">
                        <table id="paramsTable"></table>
                    </form>
                <div class="pHeaderTitleDiv top-level-background"><img src="/gp/images/toggle_collapse.png" width="19" height="19" class="paramSectionToggle">Cufflinks</div><div id="paramGroup_0_0" class="paramGroupSection solid-border paramgroup-spacing"><table class="paramsTable"><tbody><tr id="CreateSymlinks1.input.files" class="pRow"><td class="pTitle"><div class="pTitleDiv">aligned files*</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="fileDiv mainDivBorder ui-droppable" id="fileDiv-CreateSymlinks1.input.files-1"><div class="fileUploadDiv"><div class="batchBox" title="A job will be launched for every file with a matching type."><input type="checkbox" id="batchCheckCreateSymlinks1.input.files"><label for="batchCheckCreateSymlinks1.input.files">Batch</label></div><button class="uploadBtn ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" type="button" role="button" aria-disabled="false"><span class="ui-button-text">Upload Files...</span></button><div class="inputFileBtn"><input class="uploadedinputfile requiredParam" type="file" multiple="multiple"></div><button type="button" class="urlButton ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" role="button" aria-disabled="false"><span class="ui-button-text">Add Paths or URLs...</span></button><span class="drop-box">Drag Files Here</span><div class="fileSizeCaption"> 2GB file upload limit using the Upload Files... button. For files &gt; 2GB upload from the Files tab. </div></div><div class="fileListingDiv"></div></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">One or more SAM/BAM files</td></tr><tr id="CufflinksWrapperPipeline2.Cufflinks1.GTF.guide" class="pRow"><td class="pTitle"><div class="pTitleDiv">GTF guide</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="fieldTypeToggle"><div class="fileChoiceOptions ui-buttonset"><input id="selectFile_1_CufflinksWrapperPipeline2.Cufflinks1.GTF.guide" name="field_toggle_1_CufflinksWrapperPipeline2.Cufflinks1.GTF.guide" type="radio" checked="checked" class="ui-helper-hidden-accessible"><label for="selectFile_1_CufflinksWrapperPipeline2.Cufflinks1.GTF.guide" class="ui-state-active ui-button ui-widget ui-state-default ui-button-text-only ui-corner-left" role="button" aria-disabled="false"><span class="ui-button-text">Select a file</span></label><span class="elemSpacing">  or  </span><input id="customFile_1_CufflinksWrapperPipeline2.Cufflinks1.GTF.guide" name="field_toggle_1_CufflinksWrapperPipeline2.Cufflinks1.GTF.guide" type="radio" class="ui-helper-hidden-accessible"><label for="customFile_1_CufflinksWrapperPipeline2.Cufflinks1.GTF.guide" class="ui-button ui-widget ui-state-default ui-button-text-only ui-corner-right" role="button" aria-disabled="false"><span class="ui-button-text">Upload your own file</span></label> </div></div><div class="selectChoice"><select class="choice" id="CufflinksWrapperPipeline2.Cufflinks1.GTF.guide_1" style="display: none;"><option value=""></option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Arabidopsis_thaliana_TAIR10_Ensembl.gtf">Arabidopsis_thaliana_TAIR10_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Arabidopsis_thaliana_TAIR10_NCBI.gtf">Arabidopsis_thaliana_TAIR10_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Arabidopsis_thaliana_TAIR9_Ensembl.gtf">Arabidopsis_thaliana_TAIR9_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Arabidopsis_thaliana_build9_1_NCBI.gtf">Arabidopsis_thaliana_build9_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bacillus_subtilis_168_EB2_Ensembl.gtf">Bacillus_subtilis_168_EB2_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_Btau_4_0_Ensembl.gtf">Bos_taurus_Btau_4_0_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_Btau_4_6_1_NCBI.gtf">Bos_taurus_Btau_4_6_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_UMD3_1_Ensembl.gtf">Bos_taurus_UMD3_1_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_UMD_3_1_NCBI.gtf">Bos_taurus_UMD_3_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_bosTau6_UCSC.gtf">Bos_taurus_bosTau6_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_bosTau7_UCSC.gtf">Bos_taurus_bosTau7_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_WBcel215_Ensembl.gtf">Caenorhabditis_elegans_WBcel215_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_WS190_NCBI.gtf">Caenorhabditis_elegans_WS190_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_WS195_NCBI.gtf">Caenorhabditis_elegans_WS195_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_WS210_Ensembl.gtf">Caenorhabditis_elegans_WS210_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_WS220_Ensembl.gtf">Caenorhabditis_elegans_WS220_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_ce10_UCSC.gtf">Caenorhabditis_elegans_ce10_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_ce6_UCSC.gtf">Caenorhabditis_elegans_ce6_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_BROADD2_Ensembl.gtf">Canis_familiaris_BROADD2_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_CanFam3_1_Ensembl.gtf">Canis_familiaris_CanFam3_1_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_build2_1_NCBI.gtf">Canis_familiaris_build2_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_build3_1_NCBI.gtf">Canis_familiaris_build3_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_canFam2_UCSC.gtf">Canis_familiaris_canFam2_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_canFam3_UCSC.gtf">Canis_familiaris_canFam3_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Danio_rerio_Zv9_Ensembl.gtf">Danio_rerio_Zv9_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Danio_rerio_Zv9_NCBI.gtf">Danio_rerio_Zv9_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Danio_rerio_danRer7_UCSC.gtf">Danio_rerio_danRer7_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_BDGP5_25_Ensembl.gtf">Drosophila_melanogaster_BDGP5_25_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_BDGP5_Ensembl.gtf">Drosophila_melanogaster_BDGP5_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_build5_3_NCBI.gtf">Drosophila_melanogaster_build5_3_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_build5_41_NCBI.gtf">Drosophila_melanogaster_build5_41_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_build5_NCBI.gtf">Drosophila_melanogaster_build5_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_dm3_UCSC.gtf">Drosophila_melanogaster_dm3_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Equus_caballus_EquCab2_0_NCBI.gtf">Equus_caballus_EquCab2_0_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Equus_caballus_EquCab2_Ensembl.gtf">Equus_caballus_EquCab2_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Equus_caballus_equCab2_UCSC.gtf">Equus_caballus_equCab2_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Escherichia_coli_K_12_DH10B_EB1_Ensembl.gtf">Escherichia_coli_K_12_DH10B_EB1_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Gallus_gallus_WASHUC2_Ensembl.gtf">Gallus_gallus_WASHUC2_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Gallus_gallus_build2_1_NCBI.gtf">Gallus_gallus_build2_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Gallus_gallus_build3_1_NCBI.gtf">Gallus_gallus_build3_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Gallus_gallus_galGal3_UCSC.gtf">Gallus_gallus_galGal3_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Gallus_gallus_galGal4_UCSC.gtf">Gallus_gallus_galGal4_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Glycine_max_Gm01_Ensembl.gtf">Glycine_max_Gm01_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_GENCODE_hg18_v3c_chr1_format.gtf">Homo_sapiens_GENCODE_hg18_v3c_chr1_format.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_GENCODE_hg19_v7_1_format.gtf">Homo_sapiens_GENCODE_hg19_v7_1_format.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_GENCODE_hg19_v7_chr1_format.gtf">Homo_sapiens_GENCODE_hg19_v7_chr1_format.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_GRCh37_Ensembl.gtf">Homo_sapiens_GRCh37_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_build37_1_NCBI.gtf">Homo_sapiens_build37_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_build37_2_NCBI.gtf">Homo_sapiens_build37_2_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_hg18_UCSC.gtf">Homo_sapiens_hg18_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_hg19_UCSC.gtf">Homo_sapiens_hg19_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Macaca_mulatta_Mmul_1_Ensembl.gtf">Macaca_mulatta_Mmul_1_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Mus_musculus_GRCm38_Ensembl.gtf">Mus_musculus_GRCm38_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Mus_musculus_GRCm38_NCBI.gtf">Mus_musculus_GRCm38_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Mus_musculus_mm10_UCSC.gtf">Mus_musculus_mm10_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Mus_musculus_mm9_UCSC.gtf">Mus_musculus_mm9_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Oryza_sativa_japonica_MSU6_Ensembl.gtf">Oryza_sativa_japonica_MSU6_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Plasmodium_falciparum_3D7_v3_Sanger.gtf">Plasmodium_falciparum_3D7_v3_Sanger.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Rattus_norvegicus_RGSC3_4_Ensembl.gtf">Rattus_norvegicus_RGSC3_4_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Rattus_norvegicus_RGSC_v3_4_NCBI.gtf">Rattus_norvegicus_RGSC_v3_4_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Rattus_norvegicus_Rnor_5_0_NCBI.gtf">Rattus_norvegicus_Rnor_5_0_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Saccharomyces_cerevisiae_EF4_Ensembl.gtf">Saccharomyces_cerevisiae_EF4_Ensembl.gtf</option></select><button type="button" class="ui-multiselect ui-widget ui-state-default ui-corner-all mSelect" aria-haspopup="true" style="width: 315.952986650937px;"><span class="ui-icon ui-icon-triangle-1-s"></span><span></span></button></div><div class="fileDiv mainDivBorder ui-droppable" id="fileDiv-CufflinksWrapperPipeline2.Cufflinks1.GTF.guide-1" style="display: none;"><div class="fileUploadDiv"><div class="batchBox" title="A job will be launched for every file with a matching type."><input type="checkbox" id="batchCheckCufflinksWrapperPipeline2.Cufflinks1.GTF.guide"><label for="batchCheckCufflinksWrapperPipeline2.Cufflinks1.GTF.guide">Batch</label></div><button class="uploadBtn ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" type="button" role="button" aria-disabled="false"><span class="ui-button-text">Upload File...</span></button><div class="inputFileBtn"><input class="uploadedinputfile" type="file"></div><button type="button" class="urlButton ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" role="button" aria-disabled="false"><span class="ui-button-text">Add Path or URL...</span></button><span class="drop-box">Drag Files Here</span><div class="fileSizeCaption"> 2GB file upload limit using the Upload File... button. For files &gt; 2GB upload from the Files tab. </div></div><div class="fileListingDiv"></div></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">Reference annotation (GFF/GTF file) to guide RABT assembly</td></tr><tr id="CufflinksWrapperPipeline2.Cufflinks1.mask.file" class="pRow"><td class="pTitle"><div class="pTitleDiv">mask GTF file</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="fileDiv mainDivBorder ui-droppable" id="fileDiv-CufflinksWrapperPipeline2.Cufflinks1.mask.file-1"><div class="fileUploadDiv"><div class="batchBox" title="A job will be launched for every file with a matching type."><input type="checkbox" id="batchCheckCufflinksWrapperPipeline2.Cufflinks1.mask.file"><label for="batchCheckCufflinksWrapperPipeline2.Cufflinks1.mask.file">Batch</label></div><button class="uploadBtn ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" type="button" role="button" aria-disabled="false"><span class="ui-button-text">Upload File...</span></button><div class="inputFileBtn"><input class="uploadedinputfile" type="file"></div><button type="button" class="urlButton ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" role="button" aria-disabled="false"><span class="ui-button-text">Add Path or URL...</span></button><span class="drop-box">Drag Files Here</span><div class="fileSizeCaption"> 2GB file upload limit using the Upload File... button. For files &gt; 2GB upload from the Files tab. </div></div><div class="fileListingDiv"></div></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">A GTF file specifying transcripts to be ignored</td></tr><tr id="CufflinksWrapperPipeline2.Cufflinks1.library.type" class="pRow"><td class="pTitle"><div class="pTitleDiv">library type</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="selectChoice"><select class="choice" id="CufflinksWrapperPipeline2.Cufflinks1.library.type_1" style="display: none;"><option value="inferred">inferred</option><option value="fr-unstranded">fr-unstranded</option><option value="fr-firststrand">fr-firststrand</option><option value="fr-secondstrand">fr-secondstrand</option><option value="ff-unstranded">ff-unstranded</option><option value="ff-firststrand">ff-firststrand</option><option value="ff-secondstrand">ff-secondstrand</option><option value="transfrags">transfrags</option></select><button type="button" class="ui-multiselect ui-widget ui-state-default ui-corner-all mSelect" aria-haspopup="true" style="width: 224.768166691483px;"><span class="ui-icon ui-icon-triangle-1-s"></span><span>inferred</span></button></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">The library type used to generate reads.</td></tr><tr id="CufflinksWrapperPipeline2.Cufflinks1.min.frags.per.transfrag" class="pRow"><td class="pTitle"><div class="pTitleDiv">min frags per transfrag</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="textDiv"><input type="text" class="pValue"></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">Assembled transfrags supported by fewer than this many aligned RNA-Seq fragments are not reported.</td></tr></tbody></table><div class="pHeaderTitleDiv background"><img src="/gp/images/toggle_expand.png" width="19" height="19" class="paramSectionToggle">Advanced</div><div id="paramGroup_1_1" class="paramGroupSection dotted-border" style="display: none;"><table class="paramsTable"><tbody><tr id="CufflinksWrapperPipeline2.Cufflinks1.additional.cufflinks.options" class="pRow"><td class="pTitle"><div class="pTitleDiv">additional command line options</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="textDiv"><input type="text" class="pValue"></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">Additional options to be passed along to the Cufflinks program at the command line.  This parameter gives you a means to specify otherwise unavailable Cufflinks options and switches not supported by the module; check the Cufflinks manual for details.  Recommended for experts only; use this at your own discretion.</td></tr></tbody></table></div></div><div class="pHeaderTitleDiv top-level-background"><img src="/gp/images/toggle_collapse.png" width="19" height="19" class="paramSectionToggle">Cuffmerge</div><div id="paramGroup_2_0" class="paramGroupSection solid-border paramgroup-spacing"><table class="paramsTable"><tbody><tr id="Cufflinks.cuffmerge4.reference.GTF" class="pRow"><td class="pTitle"><div class="pTitleDiv">reference GTF</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="fieldTypeToggle"><div class="fileChoiceOptions ui-buttonset"><input id="selectFile_1_Cufflinks.cuffmerge4.reference.GTF" name="field_toggle_1_Cufflinks.cuffmerge4.reference.GTF" type="radio" checked="checked" class="ui-helper-hidden-accessible"><label for="selectFile_1_Cufflinks.cuffmerge4.reference.GTF" class="ui-state-active ui-button ui-widget ui-state-default ui-button-text-only ui-corner-left" role="button" aria-disabled="false"><span class="ui-button-text">Select a file</span></label><span class="elemSpacing">  or  </span><input id="customFile_1_Cufflinks.cuffmerge4.reference.GTF" name="field_toggle_1_Cufflinks.cuffmerge4.reference.GTF" type="radio" class="ui-helper-hidden-accessible"><label for="customFile_1_Cufflinks.cuffmerge4.reference.GTF" class="ui-button ui-widget ui-state-default ui-button-text-only ui-corner-right" role="button" aria-disabled="false"><span class="ui-button-text">Upload your own file</span></label> </div></div><div class="selectChoice"><select class="choice" id="Cufflinks.cuffmerge4.reference.GTF_1" style="display: none;"><option value=""></option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Arabidopsis_thaliana_TAIR10_Ensembl.gtf">Arabidopsis_thaliana_TAIR10_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Arabidopsis_thaliana_TAIR10_NCBI.gtf">Arabidopsis_thaliana_TAIR10_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Arabidopsis_thaliana_TAIR9_Ensembl.gtf">Arabidopsis_thaliana_TAIR9_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Arabidopsis_thaliana_build9_1_NCBI.gtf">Arabidopsis_thaliana_build9_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bacillus_subtilis_168_EB2_Ensembl.gtf">Bacillus_subtilis_168_EB2_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_Btau_4_0_Ensembl.gtf">Bos_taurus_Btau_4_0_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_Btau_4_6_1_NCBI.gtf">Bos_taurus_Btau_4_6_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_UMD3_1_Ensembl.gtf">Bos_taurus_UMD3_1_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_UMD_3_1_NCBI.gtf">Bos_taurus_UMD_3_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_bosTau6_UCSC.gtf">Bos_taurus_bosTau6_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Bos_taurus_bosTau7_UCSC.gtf">Bos_taurus_bosTau7_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_WBcel215_Ensembl.gtf">Caenorhabditis_elegans_WBcel215_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_WS190_NCBI.gtf">Caenorhabditis_elegans_WS190_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_WS195_NCBI.gtf">Caenorhabditis_elegans_WS195_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_WS210_Ensembl.gtf">Caenorhabditis_elegans_WS210_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_WS220_Ensembl.gtf">Caenorhabditis_elegans_WS220_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_ce10_UCSC.gtf">Caenorhabditis_elegans_ce10_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Caenorhabditis_elegans_ce6_UCSC.gtf">Caenorhabditis_elegans_ce6_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_BROADD2_Ensembl.gtf">Canis_familiaris_BROADD2_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_CanFam3_1_Ensembl.gtf">Canis_familiaris_CanFam3_1_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_build2_1_NCBI.gtf">Canis_familiaris_build2_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_build3_1_NCBI.gtf">Canis_familiaris_build3_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_canFam2_UCSC.gtf">Canis_familiaris_canFam2_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Canis_familiaris_canFam3_UCSC.gtf">Canis_familiaris_canFam3_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Danio_rerio_Zv9_Ensembl.gtf">Danio_rerio_Zv9_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Danio_rerio_Zv9_NCBI.gtf">Danio_rerio_Zv9_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Danio_rerio_danRer7_UCSC.gtf">Danio_rerio_danRer7_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_BDGP5_25_Ensembl.gtf">Drosophila_melanogaster_BDGP5_25_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_BDGP5_Ensembl.gtf">Drosophila_melanogaster_BDGP5_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_build5_3_NCBI.gtf">Drosophila_melanogaster_build5_3_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_build5_41_NCBI.gtf">Drosophila_melanogaster_build5_41_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_build5_NCBI.gtf">Drosophila_melanogaster_build5_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Drosophila_melanogaster_dm3_UCSC.gtf">Drosophila_melanogaster_dm3_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Equus_caballus_EquCab2_0_NCBI.gtf">Equus_caballus_EquCab2_0_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Equus_caballus_EquCab2_Ensembl.gtf">Equus_caballus_EquCab2_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Equus_caballus_equCab2_UCSC.gtf">Equus_caballus_equCab2_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Escherichia_coli_K_12_DH10B_EB1_Ensembl.gtf">Escherichia_coli_K_12_DH10B_EB1_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Gallus_gallus_WASHUC2_Ensembl.gtf">Gallus_gallus_WASHUC2_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Gallus_gallus_build2_1_NCBI.gtf">Gallus_gallus_build2_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Gallus_gallus_build3_1_NCBI.gtf">Gallus_gallus_build3_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Gallus_gallus_galGal3_UCSC.gtf">Gallus_gallus_galGal3_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Gallus_gallus_galGal4_UCSC.gtf">Gallus_gallus_galGal4_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Glycine_max_Gm01_Ensembl.gtf">Glycine_max_Gm01_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_GENCODE_hg18_v3c_chr1_format.gtf">Homo_sapiens_GENCODE_hg18_v3c_chr1_format.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_GENCODE_hg19_v7_1_format.gtf">Homo_sapiens_GENCODE_hg19_v7_1_format.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_GENCODE_hg19_v7_chr1_format.gtf">Homo_sapiens_GENCODE_hg19_v7_chr1_format.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_GRCh37_Ensembl.gtf">Homo_sapiens_GRCh37_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_build37_1_NCBI.gtf">Homo_sapiens_build37_1_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_build37_2_NCBI.gtf">Homo_sapiens_build37_2_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_hg18_UCSC.gtf">Homo_sapiens_hg18_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Homo_sapiens_hg19_UCSC.gtf">Homo_sapiens_hg19_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Macaca_mulatta_Mmul_1_Ensembl.gtf">Macaca_mulatta_Mmul_1_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Mus_musculus_GRCm38_Ensembl.gtf">Mus_musculus_GRCm38_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Mus_musculus_GRCm38_NCBI.gtf">Mus_musculus_GRCm38_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Mus_musculus_mm10_UCSC.gtf">Mus_musculus_mm10_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Mus_musculus_mm9_UCSC.gtf">Mus_musculus_mm9_UCSC.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Oryza_sativa_japonica_MSU6_Ensembl.gtf">Oryza_sativa_japonica_MSU6_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Plasmodium_falciparum_3D7_v3_Sanger.gtf">Plasmodium_falciparum_3D7_v3_Sanger.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Rattus_norvegicus_RGSC3_4_Ensembl.gtf">Rattus_norvegicus_RGSC3_4_Ensembl.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Rattus_norvegicus_RGSC_v3_4_NCBI.gtf">Rattus_norvegicus_RGSC_v3_4_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Rattus_norvegicus_Rnor_5_0_NCBI.gtf">Rattus_norvegicus_Rnor_5_0_NCBI.gtf</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/annotation/gtf/Saccharomyces_cerevisiae_EF4_Ensembl.gtf">Saccharomyces_cerevisiae_EF4_Ensembl.gtf</option></select><button type="button" class="ui-multiselect ui-widget ui-state-default ui-corner-all mSelect" aria-haspopup="true" style="width: 315.952986650937px;"><span class="ui-icon ui-icon-triangle-1-s"></span><span></span></button></div><div class="fileDiv mainDivBorder ui-droppable" id="fileDiv-Cufflinks.cuffmerge4.reference.GTF-1" style="display: none;"><div class="fileUploadDiv"><div class="batchBox" title="A job will be launched for every file with a matching type."><input type="checkbox" id="batchCheckCufflinks.cuffmerge4.reference.GTF"><label for="batchCheckCufflinks.cuffmerge4.reference.GTF">Batch</label></div><button class="uploadBtn ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" type="button" role="button" aria-disabled="false"><span class="ui-button-text">Upload File...</span></button><div class="inputFileBtn"><input class="uploadedinputfile" type="file"></div><button type="button" class="urlButton ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" role="button" aria-disabled="false"><span class="ui-button-text">Add Path or URL...</span></button><span class="drop-box">Drag Files Here</span><div class="fileSizeCaption"> 2GB file upload limit using the Upload File... button. For files &gt; 2GB upload from the Files tab. </div></div><div class="fileListingDiv"></div></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">An optional reference annotation GTF. The input assemblies are merged together with the reference GTF and included in the final output. Cuffmerge will use this to attach gene names and other metadata to the merged catalog.</td></tr><tr id="Cufflinks.cuffmerge4.genome.file" class="pRow"><td class="pTitle"><div class="pTitleDiv">genome file*</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="fieldTypeToggle"><div class="fileChoiceOptions ui-buttonset"><input id="selectFile_1_Cufflinks.cuffmerge4.genome.file" name="field_toggle_1_Cufflinks.cuffmerge4.genome.file" type="radio" checked="checked" class="ui-helper-hidden-accessible"><label for="selectFile_1_Cufflinks.cuffmerge4.genome.file" class="ui-state-active ui-button ui-widget ui-state-default ui-button-text-only ui-corner-left" role="button" aria-disabled="false"><span class="ui-button-text">Select a file</span></label><span class="elemSpacing">  or  </span><input id="customFile_1_Cufflinks.cuffmerge4.genome.file" name="field_toggle_1_Cufflinks.cuffmerge4.genome.file" type="radio" class="ui-helper-hidden-accessible"><label for="customFile_1_Cufflinks.cuffmerge4.genome.file" class="ui-button ui-widget ui-state-default ui-button-text-only ui-corner-right" role="button" aria-disabled="false"><span class="ui-button-text">Upload your own file</span></label> </div></div><div class="selectChoice"><select class="choice requiredParam" id="Cufflinks.cuffmerge4.genome.file_1" style="display: none;"><option value=""></option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Arabidopsis_thaliana_TAIR10_Ensembl.fa">Arabidopsis_thaliana_TAIR10_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Arabidopsis_thaliana_TAIR10_NCBI.fa">Arabidopsis_thaliana_TAIR10_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Arabidopsis_thaliana_TAIR9_Ensembl.fa">Arabidopsis_thaliana_TAIR9_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Arabidopsis_thaliana_build9_1_NCBI.fa">Arabidopsis_thaliana_build9_1_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Bacillus_cereus_ATCC_10987_2004-02-13_NCBI.fa">Bacillus_cereus_ATCC_10987_2004-02-13_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Bacillus_subtilis_168_EB2_Ensembl.fa">Bacillus_subtilis_168_EB2_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Bos_taurus_Btau_4_0_Ensembl.fa">Bos_taurus_Btau_4_0_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Bos_taurus_Btau_4_6_1_NCBI.fa">Bos_taurus_Btau_4_6_1_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Bos_taurus_UMD3_1_Ensembl.fa">Bos_taurus_UMD3_1_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Bos_taurus_UMD_3_1_NCBI.fa">Bos_taurus_UMD_3_1_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Bos_taurus_bosTau6_UCSC.fa">Bos_taurus_bosTau6_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Bos_taurus_bosTau7_UCSC.fa">Bos_taurus_bosTau7_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Caenorhabditis_elegans_WBcel215_Ensembl.fa">Caenorhabditis_elegans_WBcel215_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Caenorhabditis_elegans_WS190_NCBI.fa">Caenorhabditis_elegans_WS190_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Caenorhabditis_elegans_WS195_NCBI.fa">Caenorhabditis_elegans_WS195_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Caenorhabditis_elegans_WS210_Ensembl.fa">Caenorhabditis_elegans_WS210_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Caenorhabditis_elegans_WS220_Ensembl.fa">Caenorhabditis_elegans_WS220_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Caenorhabditis_elegans_ce10_UCSC.fa">Caenorhabditis_elegans_ce10_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Caenorhabditis_elegans_ce6_UCSC.fa">Caenorhabditis_elegans_ce6_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Canis_familiaris_BROADD2_Ensembl.fa">Canis_familiaris_BROADD2_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Canis_familiaris_CanFam3_1_Ensembl.fa">Canis_familiaris_CanFam3_1_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Canis_familiaris_build2_1_NCBI.fa">Canis_familiaris_build2_1_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Canis_familiaris_build3_1_NCBI.fa">Canis_familiaris_build3_1_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Canis_familiaris_canFam2_UCSC.fa">Canis_familiaris_canFam2_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Canis_familiaris_canFam3_UCSC.fa">Canis_familiaris_canFam3_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Danio_rerio_Zv9_Ensembl.fa">Danio_rerio_Zv9_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Danio_rerio_Zv9_NCBI.fa">Danio_rerio_Zv9_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Danio_rerio_danRer7_UCSC.fa">Danio_rerio_danRer7_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Drosophila_melanogaster_BDGP5_25_Ensembl.fa">Drosophila_melanogaster_BDGP5_25_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Drosophila_melanogaster_BDGP5_Ensembl.fa">Drosophila_melanogaster_BDGP5_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Drosophila_melanogaster_build5_3_NCBI.fa">Drosophila_melanogaster_build5_3_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Drosophila_melanogaster_build5_41_NCBI.fa">Drosophila_melanogaster_build5_41_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Drosophila_melanogaster_build5_NCBI.fa">Drosophila_melanogaster_build5_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Drosophila_melanogaster_dm3_UCSC.fa">Drosophila_melanogaster_dm3_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Enterobacteriophage_lambda_1993-04-28_NCBI.fa">Enterobacteriophage_lambda_1993-04-28_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Equus_caballus_EquCab2_0_NCBI.fa">Equus_caballus_EquCab2_0_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Equus_caballus_EquCab2_Ensembl.fa">Equus_caballus_EquCab2_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Equus_caballus_equCab2_UCSC.fa">Equus_caballus_equCab2_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Escherichia_coli_K_12_DH10B_2008-03-17_NCBI.fa">Escherichia_coli_K_12_DH10B_2008-03-17_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Escherichia_coli_K_12_DH10B_EB1_Ensembl.fa">Escherichia_coli_K_12_DH10B_EB1_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Escherichia_coli_K_12_MG1655_2001-10-15_NCBI.fa">Escherichia_coli_K_12_MG1655_2001-10-15_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Gallus_gallus_WASHUC2_Ensembl.fa">Gallus_gallus_WASHUC2_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Gallus_gallus_build2_1_NCBI.fa">Gallus_gallus_build2_1_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Gallus_gallus_build3_1_NCBI.fa">Gallus_gallus_build3_1_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Gallus_gallus_galGal3_UCSC.fa">Gallus_gallus_galGal3_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Gallus_gallus_galGal4_UCSC.fa">Gallus_gallus_galGal4_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Glycine_max_Gm01_Ensembl.fa">Glycine_max_Gm01_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Homo_sapiens_GRCh37_Ensembl.fa">Homo_sapiens_GRCh37_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Homo_sapiens_build37_1_NCBI.fa">Homo_sapiens_build37_1_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Homo_sapiens_build37_2_NCBI.fa">Homo_sapiens_build37_2_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Homo_sapiens_hg18_UCSC.fa">Homo_sapiens_hg18_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Homo_sapiens_hg19_UCSC.fa">Homo_sapiens_hg19_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Macaca_mulatta_Mmul_1_Ensembl.fa">Macaca_mulatta_Mmul_1_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Mus_musculus_GRCm38_Ensembl.fa">Mus_musculus_GRCm38_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Mus_musculus_GRCm38_NCBI.fa">Mus_musculus_GRCm38_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Mus_musculus_mm10_UCSC.fa">Mus_musculus_mm10_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Mus_musculus_mm9_UCSC.fa">Mus_musculus_mm9_UCSC.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Mycobacterium_tuberculosis_H37RV_2001-09-07_NCBI.fa">Mycobacterium_tuberculosis_H37RV_2001-09-07_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Mycobacterium_tuberculosis_H37RV_H37Rv_EB1_Ensembl.fa">Mycobacterium_tuberculosis_H37RV_H37Rv_EB1_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Oryza_sativa_japonica_MSU6_Ensembl.fa">Oryza_sativa_japonica_MSU6_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Plasmodium_falciparum_3D7_v3_Sanger.fa">Plasmodium_falciparum_3D7_v3_Sanger.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Pseudomonas_aeruginosa_PAO1_2000-09-13_NCBI.fa">Pseudomonas_aeruginosa_PAO1_2000-09-13_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Rattus_norvegicus_RGSC3_4_Ensembl.fa">Rattus_norvegicus_RGSC3_4_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Rattus_norvegicus_RGSC_v3_4_NCBI.fa">Rattus_norvegicus_RGSC_v3_4_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Rattus_norvegicus_Rnor_5_0_NCBI.fa">Rattus_norvegicus_Rnor_5_0_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Saccharomyces_cerevisiae_EF4_Ensembl.fa">Saccharomyces_cerevisiae_EF4_Ensembl.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Saccharomyces_cerevisiae_build3_1_NCBI.fa">Saccharomyces_cerevisiae_build3_1_NCBI.fa</option><option value="ftp://gpftp.broadinstitute.org/module_support_files/sequence/whole_genome/Saccharomyces_cerevisiae_sacCer3_UCSC.fa">Saccharomyces_cerevisiae_sacCer3_UCSC.fa</option></select><button type="button" class="ui-multiselect ui-widget ui-state-default ui-corner-all mSelect" aria-haspopup="true" style="width: 354px;"><span class="ui-icon ui-icon-triangle-1-s"></span><span></span></button></div><div class="fileDiv mainDivBorder ui-droppable" id="fileDiv-Cufflinks.cuffmerge4.genome.file-1" style="display: none;"><div class="fileUploadDiv"><div class="batchBox" title="A job will be launched for every file with a matching type."><input type="checkbox" id="batchCheckCufflinks.cuffmerge4.genome.file"><label for="batchCheckCufflinks.cuffmerge4.genome.file">Batch</label></div><button class="uploadBtn ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" type="button" role="button" aria-disabled="false"><span class="ui-button-text">Upload File...</span></button><div class="inputFileBtn"><input class="uploadedinputfile requiredParam" type="file"></div><button type="button" class="urlButton ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" role="button" aria-disabled="false"><span class="ui-button-text">Add Path or URL...</span></button><span class="drop-box">Drag Files Here</span><div class="fileSizeCaption"> 2GB file upload limit using the Upload File... button. For files &gt; 2GB upload from the Files tab. </div></div><div class="fileListingDiv"></div></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">A file containing the genomic DNA sequences for the reference.  This should be a multi-FASTA file with all contigs present.</td></tr></tbody></table></div><div class="pHeaderTitleDiv top-level-background"><img src="/gp/images/toggle_expand.png" width="19" height="19" class="paramSectionToggle">Lsf Flags</div><div id="paramGroup_3_0" class="paramGroupSection solid-border paramgroup-spacing" style="display: none;"><div class="pHeaderDescription">Optional LSF job resource requirements. These options correspond to command line args to the 'bsub' command.</div><table class="paramsTable"><tbody><tr id="job.queue" class="pRow"><td class="pTitle"><div class="pTitleDiv">queue</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="selectChoice"><select class="choice" id="job.queue_1" style="display: none;"><option value="genepattern">genepattern</option><option value="hour">hour</option><option value="week">week</option><option value="preview">preview</option></select><button type="button" class="ui-multiselect ui-widget ui-state-default ui-corner-all mSelect" aria-haspopup="true" style="width: 199.025307642265px;"><span class="ui-icon ui-icon-triangle-1-s"></span><span>genepattern</span></button></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">All jobs are run as the 'gpdev' role account, by default on the 'genepattern' queue. See the BITS Wiki (https://it.broadinstitute.org/wiki/LSF) for details on the other queues.</td></tr><tr id="job.memory" class="pRow"><td class="pTitle"><div class="pTitleDiv">memory</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="selectChoice"><select class="choice" id="job.memory_1" style="display: none;"><option value="1">small (1 Gb)</option><option value="2">default (2 Gb)</option><option value="8">8 Gb</option><option value="12">12 Gb</option><option value="16">16 Gb</option><option value="24">max available (24 Gb)</option><option value="36">more than max (32 Gb)</option><option value="64">pending indefinitely (64 Gb)</option></select><button type="button" class="ui-multiselect ui-widget ui-state-default ui-corner-all mSelect" aria-haspopup="true" style="width: 276.572974344542px;"><span class="ui-icon ui-icon-triangle-1-s"></span><span>default (2 Gb)</span></button></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">An integer, in number of Gb, default is '2'</td></tr><tr id="lsf.cpu.slots" class="pRow"><td class="pTitle"><div class="pTitleDiv">cpu slots</div></td><td class="paramValueTd"><div class="valueEntryDiv"><div class="selectChoice"><select class="choice" id="lsf.cpu.slots_1" style="display: none;"><option value="1">1</option><option value="2">2</option><option value="4">4</option><option value="6">6</option></select><button type="button" class="ui-multiselect ui-widget ui-state-default ui-corner-all mSelect" aria-haspopup="true" style="width: 230.124863945902px;"><span class="ui-icon ui-icon-triangle-1-s"></span><span>1</span></button></div></div></td></tr><tr class="paramDescription"><td></td><td colspan="3">The '-n' arg to the bsub command</td></tr></tbody></table></div></div>

                <div class="pHeaderTitleDiv top-level-background">
                    <img src="/gp/images/toggle_collapse.png" alt="toggle image" width="19" height="19" class="paramSectionToggle">
                    Comments
                </div>
                <div class="commentsContent paramgroup-spacing">
                    <textarea id="jobComment" placeholder="Add a comment..."></textarea>
                </div>

               <div class="pHeaderTitleDiv top-level-background">
                    <img src="/gp/images/toggle_collapse.png" alt="toggle image" width="19" height="19" class="paramSectionToggle">
                    Tags
                </div>
                <div class="tagsContent paramgroup-spacing">
                    <input id="tags1425055056087" style="display: none;"><div id="tags1425055056087_tagsinput" class="tagsinput" style="width: 97%; min-height: 40px; height: 100%;"><div id="tags1425055056087_addTag"><input id="tags1425055056087_tag" value="" data-default="Add tag and press enter..." class="ui-autocomplete-input" autocomplete="off" style="color: rgb(204, 204, 204); width: 80px;"><span role="status" aria-live="polite" class="ui-helper-hidden-accessible"></span></div><div class="tags_clear"></div></div>
                </div>

                <hr>

                <div style="height: 40px;">
                <span class="floatLeft otherControlsDiv">
                </span>
                <span class="floatRight submitControlsDiv">
                    <button class="Reset ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" role="button" aria-disabled="false"><span class="ui-button-text"><img src="/gp/images/reset.gif" width="16" height="16" alt="Reset" class="buttonIcon"> Reset </span></button>
                    <button class="Run ui-button ui-widget ui-state-default ui-corner-all ui-button-text-only" role="button" aria-disabled="false"><span class="ui-button-text"><img src="/gp/images/run-green.gif" width="16" height="16" alt="Run" class="buttonIcon"> Run </span></button>
                </span>
                </div>
            </div>